pkg/scaling/scale_handler.go

```diff
@@ -476,62 +489,76 @@ func (h *scaleHandler) GetScaledObjectMetrics(ctx context.Context, scaledObjectN
+				wg.Add(1)
+				go func(results chan metricResult, wg *sync.WaitGroup, metricName string, scalerIndex int, scalerConfig scalers.ScalerConfig, spec v2.MetricSpec) {
+					result := metricResult{}
+
+					// Pair metric values with its trigger names. This is applied only when
+					// ScalingModifiers.Formula is defined in SO.
+					result.metricTriggerPair, err = modifiers.GetPairTriggerAndMetric(scaledObject, metricName, scalerConfig.TriggerName)
+					if err != nil {
+						logger.Error(err, "error pairing triggers & metrics for compositeScaler")
+					}
+					var metrics []external_metrics.ExternalMetricValue
+
+					// if cache is defined for this scaler/metric, let's try to hit it first
+					metricsFoundInCache := false
+					if scalerConfig.TriggerUseCachedMetrics {
+						var metricsRecord metricscache.MetricsRecord
+						if metricsRecord, metricsFoundInCache = h.scaledObjectsMetricCache.ReadRecord(scaledObjectIdentifier, metricName); metricsFoundInCache {
+							logger.V(1).Info("Reading metrics from cache", "scaler", scalerName, "metricName", metricName, "metricsRecord", metricsRecord)
+							metrics = metricsRecord.Metric
+							err = metricsRecord.ScalerError
+						}
 					}
-				}
 
-				if !metricsFoundInCache {
-					var latency int64
-					metrics, _, latency, err = cache.GetMetricsAndActivityForScaler(ctx, scalerIndex, metricName)
-					if latency != -1 {
-						metricscollector.RecordScalerLatency(scaledObjectNamespace, scaledObject.Name, scalerName, scalerIndex, metricName, float64(latency))
+					if !metricsFoundInCache {
+						var latency int64
+						metrics, _, latency, err = cache.GetMetricsAndActivityForScaler(ctx, scalerIndex, metricName)
+						if latency != -1 {
+							metricscollector.RecordScalerLatency(scaledObjectNamespace, scaledObject.Name, scalerName, scalerIndex, metricName, float64(latency))
+						}
+						logger.V(1).Info("Getting metrics from scaler", "scaler", scalerName, "metricName", metricName, "metrics", metrics, "scalerError", err)
 					}
-					logger.V(1).Info("Getting metrics from scaler", "scaler", scalerName, "metricName", spec.External.Metric.Name, "metrics", metrics, "scalerError", err)
-				}
 
-				// check if we need to set a fallback
-				metrics, fallbackApplied, err := fallback.GetMetricsWithFallback(ctx, h.client, metrics, err, metricName, scaledObject, spec)
-				if fallbackApplied {
-					fallbackActive = true
-				}
-				if err != nil {
-					isScalerError = true
-					logger.Error(err, "error getting metric for scaler", "scaler", scalerName)
-				} else {
-					for _, metric := range metrics {
-						metricValue := metric.Value.AsApproximateFloat64()
-						metricscollector.RecordScalerMetric(scaledObjectNamespace, scaledObjectName, scalerName, scalerIndex, metric.MetricName, metricValue)
+					// check if we need to set a fallback
+					result.metrics, result.fallbackActive, err = fallback.GetMetricsWithFallback(ctx, h.client, metrics, err, metricName, scaledObject, spec)
```

Comment:
down in the execution path, this will call kube-api with status update with `PATCH` on the same `ScaledObject`. I think it would be safer now to aggregate the status conditions and then call the `PATCH` by the end with all changes. These quick bursts of patches might get rate-limited given they are all triggered at the same time in goroutines.

Reply:
It's a good point, I'll take a look later on ❤️

Reply:
+1

Reply:
PTAL